OpenRA.Game/Graphics/SpriteFont.cs

```diff
@@ -233,6 +233,52 @@ public void DrawTextWithShadow(string text, float2 location, Color fg, Color bgD
 			DrawTextWithShadow(text, location, fg, GetContrastColor(fg, bgDark, bgLight), offset, angle);
 		}
 
+		public void DrawTextWithSelection(string text, float2 location, Color normal, Color selected, Color selectedBg, int selectionStart, int selectionEnd)
+		{
+			var currentIndex = 0;
+
+			// Offset from the baseline position to the top-left of the glyph for rendering
+			var rowLocation = location + new float2(0, size);
+			var columnLocation = rowLocation;
+
+			foreach (var character in text)
+			{
+				if (character == '\n')
+				{
+					rowLocation += new float2(0, size);
+					columnLocation = rowLocation;
+					continue;
+				}
+
+				var glyph = glyphs[character];
+				var inSelection = selectionStart <= currentIndex && currentIndex <= selectionEnd;
+				var tintColor = inSelection ? selected : normal;
+				var tint = new float3(tintColor.R / 256f, tintColor.G / 256f, tintColor.B / 256f);
+
+				if (inSelection)
+				{
+					// draw highlights from the top of the line, not the baseline.
+					Game.Renderer.RgbaColorRenderer.FillRect(
```

Comment:
There are two issues with doing this character by character:
* It's really inefficient
* If the text flows over multiple lines the selection background for lower lines hides the tails of long letters like `j`.

I think it would be better to avoid adding this new method (at least until we have more use cases), and instead have `LabelWithSelectionWidget` measure the width of the selected text in each line and draw a single rectangle for each line (and all lines) before using the normal DrawText on top.